server/src/main/java/org/elasticsearch/snapshots/SnapshotInfo.java

```diff
@@ -370,23 +373,30 @@ public SnapshotInfo(
     }
 
     public SnapshotInfo(SnapshotsInProgress.Entry entry) {
-        this(
-            entry.snapshot(),
-            List.copyOf(entry.indices().keySet()),
-            entry.dataStreams(),
-            entry.featureStates(),
-            null,
-            Version.CURRENT,
-            entry.startTime(),
-            0L,
-            0,
-            0,
-            Collections.emptyList(),
-            entry.includeGlobalState(),
-            entry.userMetadata(),
-            SnapshotState.IN_PROGRESS,
-            Collections.emptyMap()
-        );
+        int successfulShards = 0;
+        List<SnapshotShardFailure> shardFailures = new ArrayList<>();
+        for (ObjectObjectCursor<ShardId, SnapshotsInProgress.ShardSnapshotStatus> c : entry.shards()) {
+            if (c.value.state() == SnapshotsInProgress.ShardState.SUCCESS) {
+                successfulShards++;
+            } else if (c.value.state() == SnapshotsInProgress.ShardState.FAILED) {
+                shardFailures.add(new SnapshotShardFailure(c.value.nodeId(), c.key, c.value.reason()));
+            }
+        }
+        this.snapshot = Objects.requireNonNull(entry.snapshot());
+        this.indices = List.copyOf(entry.indices().keySet());
+        this.dataStreams = List.copyOf(entry.dataStreams());
+        this.featureStates = List.copyOf(entry.featureStates());
+        this.state = SnapshotState.IN_PROGRESS;
+        this.reason = null;
+        this.version = Version.CURRENT;
+        this.startTime = entry.startTime();
+        this.endTime = 0L;
+        this.totalShards = entry.shards().size();
+        this.successfulShards = successfulShards;
+        this.shardFailures = Collections.unmodifiableList(shardFailures);
```

Comment:
I think we can compute the shardFailures once in a static method and then compute `successfulShards = entry.shards().size()-shardFailures.size()` ?

Reply:
I believe it doesn't work for snapshots in progress, because not all shards that didn't fail were successful. There might be shards in the `INIT`, `WAITING`, or `QUEUED` states which were not processed yet.

Reply:
Of course 🤦🏻